.github/workflows/lint-test-build-push.yaml

```diff
@@ -0,0 +1,207 @@
+name: Lint, Test, Build, Push
+
+on:
+  push:
+    branches:
+      - master
+  pull_request:
+
+
+jobs:
+  lint-test-build-push:
+    runs-on: ubuntu-latest
+    env:
+      # Determine whether or not we should build the
+      # final production image and push it to GHCR.
+      production_build: ${{ github.event_name != 'pull_request' &&
+        github.ref == 'refs/heads/sebastiaan/backend/cache-docker-images' }}
+
+    steps:
+      # Create a short SHA-tag to tag built images
+      - name: Create SHA Container Tag
+        id: sha_tag
+        run: |
+          tag=$(cut -c 1-7 <<< $GITHUB_SHA)
+          echo "::set-output name=tag::$tag"
+      - name: Checkout code
+        uses: actions/checkout@v2
+
+      - name: Set up Docker Buildx
+        uses: docker/setup-buildx-action@v1
+
+      - name: Login to Github Container Registry
+        uses: docker/login-action@v1
+        with:
+          registry: ghcr.io
+          username: ${{ github.repository_owner }}
+          password: ${{ secrets.GHCR_TOKEN  }}
+
+      # Set up a caching directory for image layers. According to the docker
+      # documentation, it's recommended to use a SHA-based key to get the
+      # greatest change of finding the most relevant cached layer. We fall
+      # down to more generic containers by then matching by GitHub branch,
+      # to use cache generated earlier in the same branch, and finally to
+      # the latest cache in general. The `v0` is purely a cache version
+      # indicator that can be incremented manually if we want to invalidate
+      # old caches completely.
+      - name: Cache Image Layers
+        uses: actions/cache@v2
+        with:
+          path: /tmp/.buildx-cache
+          key: ${{ runner.os }}-v0-buildx-${{ github.ref }}-${{ github.sha }}
+          restore-keys: |
+            ${{ runner.os }}-v0-buildx-${{ github.ref }}-
+            ${{ runner.os }}-v0-buildx-
```

Comment:
If it always restores a cache, doesn't this mean the cache will get bloated with old layers as time goes on?

Reply:
I was wondering the same thing, but I'm following a [recipe documented in the README for the docker action](https://github.com/docker/build-push-action#leverage-github-cache). I've tried looking for more information on this, but I haven't found it yet.

Reply:
I think what is restored from the partial cache hit is written to the new cache created when the workflow ends, since the cache action just saves whatever is in a particular directory. Maybe running `docker system prune -f` before ending the workflow will avoid this problem. Hopefully that will not clear stuff related to the images that were just built. However, I think the `-a` option might (how would it know which images are "in use").

Reply:
I'm not entirely sure if that will work for a `type=local` cache, going by [this issue](https://github.com/docker/buildx/issues/310). Currently, I'm thinking of running this with a repository-only cache to see how that affects build speeds. If the difference in build speed is negligible, maybe we can just do away with local caching entire and rely on GHCR for our layer cache.

Reply:
That seems to work well. I think it may lead to faster builds in some situations and longer builds in some others, where the `latest` image is a bit out of sync with cache-invalidating changes made in a feature branch. However, if that's not the cache, this seems to be faster than using the local cache option.

A fully non-cached build took 3m7s; a fully cached build took 1m45s, using just the repository cache. I think we should just be happy with that and not worry about the local cache that could potentially get bloated.

If we notice systematically longer builds in the future, we could try another approach (e.g, making the local cache branch specific and manually creating a clean cache dir once per x days for master).

Reply:
One thing I forgot is that when it restores a cache from master, the new cache it writes will be scoped locally rather to master. Therefore, PRs will only bloat caches for PRs. Since PRs don't tend to have very long lifetimes, bloating shouldn't be as bad. It's still not ideal, since every new commit a PR makes will add new layers to the cache without getting rid of layers specific to previous commits. However, I imagine the layers will be small (basically just the source code copy step) unless every commit is constantly changing the venv or base layers.

I think the local cache should be disabled on master because the bloat could live on indefinitely, assuming enough activity that the cache doesn't expire. A local cache isn't generally useful on master for two reasons:

1. I don't believe the final merge to master has access to the PR's cache scope. 
2. If the previous run on master succeeded, then the images will be available from the registry, making the cache redundant.